README.md
# hyphe-lib-ger.js

[![npm version](https://badge.fury.io/js/hyphe-lib-ger.svg)](https://badge.fury.io/js/hyphe-lib-ger) [![Build Status](https://travis-ci.org/blurrryy/hyphe-lib-ger.svg?branch=master)](https://travis-ci.org/blurrryy/hyphe-lib-ger) [![](https://data.jsdelivr.com/v1/package/npm/hyphe-lib-ger/badge)](https://www.jsdelivr.com/package/npm/hyphe-lib-ger)

## Add soft-hyphens to German text

[CHANGELOG](CHANGELOG.md)

### About

Using the packages [hypher](https://github.com/bramstein/hypher)
and [hyphenation.de](https://github.com/bramstein/hyphenation-patterns) to parse a normal, German text adding soft-hyphens to allow soft-hyphens in older browsers and Google Chrome.

New: Use [typographizer-js](https://github.com/ovlb/typographizer-js) to produce better looking typography

### Install

In your terminal use

```bash
npm i --save hyphe-lib-ger
```

In your JavaScript file use

```js
const hyphe = require("hyphe-lib-ger");
```

### Usage

```js
hyphe(Text, { Options }).then(text => {
  console.log(text); // Output the formatted text
});
```

Text: String
A String with the text to parse

Options: Object
The possible options (not required)

### Options

typographize: Boolean (default: false)

- parse the text with typograhize-js first

ignoreLineBreaks: Boolean (default: false)

- ignore line breaks completely?

escapeToHTML: Boolean (default: true)

- escape all non-word characters to their HTML charcode alternative

### Example

```js
const hyphe = require("hyphe-lib-ger");

hyphe(
  `Lass mich dir kurz vorstellen, was hier eigentlich möglich ist: "Hallo, mein Name ist Daniel"`
).then(res => {
  console.log(res);
});

/*
Will return:
Lass&#32;mich&#32;dir&#32;kurz&#32;vor&#173;stel&#173;len&#44;&#32;was&#32;hier&#32;ei&#173;gent&#173;lich&#32;mög&#173;lich&#32;ist&#58;&#32;&#8220;Hal&#173;lo&#44;&#32;mein&#32;Na&#173;me&#32;ist&#32;Da&#173;ni&#173;el&#8221;
*/
```

### Example

A working example with this lib and Angular6 is seen [here](https://blurrryy.github.io)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [3.0.2] - 2019-11-12

### Fixed

- Fixed syntax errors while linting

## [3.0.1] - 2019-11-12

### Added

- Added integration testing for the lib
- Browser package has returned using HypheLib()

### Fixed

- Passing no string now returns an empty string
- Passing and empty string now returns an empty string
- A single options no longer overwrites the other default options
- An unknown option will now be ignored
- Multiple hyphe requests with different options will now work seperately

## [3.0.0] - 2019-11-05

### Added

- Add the typographizer-js package to produce better typography

### Changed

- The package now returns a promise to support typographizer
- Reworked the code to get better results from non word characters
- The options have been completely changed
- typographize? Wheter or not to use the typographizer-js package
- ignoreLineBreaks? Ignore all line breaks in the string
- The default mode now supports unicode
- escapeToHTML? Switch to escape every soft-hyphen and non word character to it's HTML char code representation

## [2.2.0] - 2018-06-26

### Added

- Add the options to add zero-width spaces after a forward slash (see README)

## [2.1.1] - 2018-06-23

### Changed

- Fixed some typos in the documentation
- Slightly changed README
- Add Browser-Endpoint to package.json

## [2.1.0] - 2018-06-23

### Added

- Add webpack version to the npm package (see README)

### Changed

- Minor fixes in versioning
- Fix some spelling issues
- Change to webpack for browserfied version

## [2.0.1] - 2018-06-22

### Changed

- Minor Bugfixes due to publishing the package

## [2.0.0] - 2018-06-22

### Added

- The script now correctly parses linebreaks
- You can now completely ignore linebreaks
- You can set a custom string as linebreak in HTML mode

### Changed

- Code completely rewritten
- Options are now handled in a different way
- Readme now shows the new usage of the library

### Removed

- Removed usage of callbacks

## [1.0.2] - 2017-09-25

### Changed

- Fix a bug where numbers wheren't parsed

## [1.0.1] - 2017-09-22

### Added

- CHANGELOG

### Changed

- README now shows the correct installation instructions

## [1.0.0] - 2017-09-21

### Added

- Parser can output in Unicode now

### Changed

- Dramatically better parser using RegEx
- Now parsing all words in a correct way
- Update README to make usage production-able

## [0.1.1] - 2017-09-21

### Changed

- Clean repo for publishing as npm package
- Beautify the code


lib/hyphe-lib-ger.spec.js
const hyphe = require("./hyphe-lib-ger");

test("Return an empty string when passing no arguments", done => {
  hyphe().then(string => {
    expect(string).toBe(``);
    done();
  });
});

test("Return an empty string when passing an empty string with default options", done => {
  const testString = "";
  hyphe(testString).then(string => {
    expect(string).toBe(``);
    done();
  });
});

test("Output with default options", done => {
  const testString = "Hello World";
  hyphe(testString).then(string => {
    expect(string).toBe(`Hel&#173;lo&#32;World`);
    done();
  });
});

test("Output should ignore unknown options", done => {
  const testString = "Hello World";
  hyphe(testString, { giveMeAnError: true }).then(string => {
    expect(string).toBe(`Hel&#173;lo&#32;World`);
    done();
  });
});

test("Given an option should not ignore the other default options", done => {
  const testString = "Hello World";
  hyphe(testString, { ignoreLineBreaks: true }).then(string => {
    expect(string).toBe(`Hel&#173;lo&#32;World`);
    done();
  });
});

test("Correct use of typographizer-js and default options", done => {
  const testString = "Ich fragte: 'Wie geht es dir?'";
  hyphe(testString).then(string => {
    expect(string).toBe(
      `Ich&#32;frag&#173;te&#58;&#32;&#8216;Wie&#32;geht&#32;es&#32;dir&#63;&#39;`
    );
    done();
  });
});

test("Correct use of typographizer-js and without HTML output", done => {
  const testString = `Ich fragte: "Wie geht es dir?"`;
  hyphe(testString, { escapeToHTML: false }).then(string => {
    expect(string).toBe(`Ich frag­te: "Wie geht es dir?"`);
    done();
  });
});


lib/hyphe-lib-ger-v3.js
const hypher = require("hypher");
const langDe = require("hyphenation.de");
const TypograhizerJS = require("typographizer-js");

const hypherGer = new hypher(langDe);
const typo = new TypograhizerJS();

const defaultOpts = {
  typographize: true,
  ignoreLineBreaks: false,
  escapeToHTML: true
};

const hypheLib = {
  __softHypen: String.fromCharCode(173),
  __regExSplit: /[\wöäüÖÜÄß]+|\W/g,
  __opts: {},
  async typographize(string) {
    return await typo.typographize(string);
  },
  escapeToHTML(word) {
    if (this.__opts.escapeToHTML) {
      const old = word;
      word = "";
      for (let i in old) {
        word += `&#${old.charCodeAt(i)};`;
      }
    }
    return word;
  },
  applyToSingleWord(word) {
    return hypherGer.hyphenate(word).join(this.escapeToHTML(this.__softHypen));
  },
  applyToNonWordChars(char) {
    let charCode = char[0].charCodeAt(0);
    switch (charCode) {
      case 47:
        return this.__opts.ignoreLineBreaks
          ? char
          : this.escapeToHTML(char + String.fromCharCode(8203));
      default:
        return this.escapeToHTML(char);
    }
  },
  async run(string, opts) {
    this.__opts = opts;
    if (opts.typographize) string = await this.typographize(string);
    string = string.match(this.__regExSplit).map(word => {
      if (word.match(/[^\wöäüÖÜÄß]/g)) return this.applyToNonWordChars(word);
      else return this.applyToSingleWord(word);
    });
    return string.join("");
  }
};

module.exports = (string, opts = defaultOpts) => {
  return new Promise(async (resolve, reject) => {
    try {
      const output = await hypheLib.run(string, opts);
      resolve(output);
    } catch (err) {
      reject(err);
    }
  });
};


lib/hyphe-lib-ger.js
const hypher = require("hypher");
const langDe = require("hyphenation.de");
const TypograhizerJS = require("typographizer-js");

const hypherGer = new hypher(langDe);
const typo = new TypograhizerJS();

const defaultOpts = {
  typographize: true,
  ignoreLineBreaks: false,
  escapeToHTML: true
};

const hypheLib = () => {
  return {
    __softHypen: String.fromCharCode(173),
    __regExSplit: /[\wöäüÖÜÄß]+|\W/g,
    __opts: {},
    async typographize(string) {
      return await typo.typographize(string);
    },
    escapeToHTML(word) {
      if (this.__opts.escapeToHTML) {
        const old = word;
        word = "";
        for (let i in old) {
          word += `&#${old.charCodeAt(i)};`;
        }
      }
      return word;
    },
    applyToSingleWord(word) {
      return hypherGer
        .hyphenate(word)
        .join(this.escapeToHTML(this.__softHypen));
    },
    applyToNonWordChars(char) {
      let charCode = char[0].charCodeAt(0);
      switch (charCode) {
        case 47:
          return this.__opts.ignoreLineBreaks
            ? char
            : this.escapeToHTML(char + String.fromCharCode(8203));
        default:
          return this.escapeToHTML(char);
      }
    },
    async run(string, opts) {
      this.__opts = { ...defaultOpts, ...opts };
      if (!string || string.length < 1) return "";
      if (opts.typographize) string = await this.typographize(string);
      string = string.match(this.__regExSplit).map(word => {
        if (word.match(/[^\wöäüÖÜÄß]/g)) return this.applyToNonWordChars(word);
        else return this.applyToSingleWord(word);
      });
      return string.join("");
    }
  };
};

module.exports = (string, opts = defaultOpts) => {
  return new Promise((resolve, reject) => {
    hypheLib()
      .run(string, opts)
      .then(output => {
        resolve(output);
      })
      .catch(err => reject(err));
  });
};
